base/inference.jl

```diff
@@ -1489,94 +1451,79 @@ function newvar!(sv::InferenceState, typ)
     return GenSym(id)
 end
 
-# copy a LambdaInfo just enough to make it not share data with li.def
-function unshare_linfo(li::LambdaInfo)
-    if li.nargs > 0
-        if li === li.def
-            li = ccall(:jl_copy_lambda_info, Any, (Any,), li)::LambdaInfo
-        end
-        if !isa(li.code, Array{Any,1})
-            li.code = ccall(:jl_uncompress_ast, Any, (Any,Any), li, li.code)
-        elseif li.code === li.def.code && li !== li.def
-            li.code = astcopy(li.code)
-        end
-        li.slotnames = copy(li.slotnames)
-        li.slotflags = copy(li.slotflags)
-        if isa(li.slottypes,Array)
-            li.slottypes = copy(li.slottypes)
-        end
-        if isa(li.gensymtypes,Array)
-            li.gensymtypes = copy(li.gensymtypes)
-        end
+# create a specialized LambdaInfo from a method
+function specialize_method(method::Method, types::ANY, sp::SimpleVector)
+    li = ccall(:jl_get_specialized, Any, (Any, Any, Any), method, types, sp)::LambdaInfo
+end
+
+# create copies of any field that type-inference might modify
+function unshare_linfo!(li::LambdaInfo)
+    if !isa(li.code, Array{Any,1})
+        li.code = ccall(:jl_uncompress_ast, Any, (Any,Any), li, li.code)
+    else
+        li.code = astcopy(li.code)
     end
-    if !isa(li.slottypes,Array)
-        li.slottypes = Any[ Any for i = 1:length(li.slotnames) ]
+    li.slotnames = copy(li.slotnames)
+    li.slotflags = copy(li.slotflags)
+    if isa(li.slottypes, Array)
+        li.slottypes = copy(li.slottypes)
     end
-    if !isa(li.gensymtypes,Array)
-        li.gensymtypes = Any[ NF for i = 1:(li.gensymtypes::Int) ]
+    if isa(li.gensymtypes, Array)
+        li.gensymtypes = copy(li.gensymtypes)
     end
     return li
 end
 
 #### entry points for inferring a LambdaInfo given a type signature ####
 
-function typeinf_edge(linfo::LambdaInfo, atypes::ANY, sparams::SimpleVector, needtree::Bool, optimize::Bool, cached::Bool, caller)
-    #println(linfo)
-    if linfo.module === Core && isempty(sparams) && isempty(linfo.sparam_vals)
+function typeinf_edge(method::Method, atypes::ANY, sparams::SimpleVector, needtree::Bool, optimize::Bool, cached::Bool, caller)
+    #println(method)
+    if method.module === Core && isempty(method.lambda_template.sparam_syms)
         atypes = Tuple
     end
-    local tfunc_idx = -1
-    local frame
+    local frame = nothing
+    offs = 0
```

Comment:
compatibility with the existing API inherited from the previous implementation of `MethodTable->cache_targs / cache_args1`. I would like to drop `offs` entirely (I have a commit to do it because it enables another optimization, but it only seemed to save ~0.5% on yichao's dispatch perf test), but in the existing codebase, it is used to decide whether the cache should split on the first or second argument type.